test/unit/content-src/lib/selectLayoutRender.test.js

```diff
@@ -87,12 +122,17 @@ describe("selectLayoutRender", () => {
     store.dispatch({type: at.DISCOVERY_STREAM_SPOCS_UPDATE, data: fakeSpocsData});
     const randomStub = globals.sandbox.stub(global.Math, "random").returns(0.6);
 
-    const result = selectLayoutRender(store.getState().DiscoveryStream, {}, []);
+    const {spocsFill, layoutRender} = selectLayoutRender(store.getState().DiscoveryStream, {}, []);
 
     assert.calledTwice(randomStub);
-    assert.lengthOf(result, 1);
-    assert.deepEqual(result[0].components[0].data.recommendations[0], "foo");
-    assert.deepEqual(result[0].components[0].data.recommendations[1], "bar");
+    assert.lengthOf(layoutRender, 1);
+    assert.deepEqual(layoutRender[0].components[0].data.recommendations[0], "foo");
+    assert.deepEqual(layoutRender[0].components[0].data.recommendations[1], "bar");
+
+    assert.deepEqual(spocsFill, [
+      {id: undefined, reason: "probability_selection", displayed: 0, full_recalc: 0},
+      {id: undefined, reason: "out_of_position", displayed: 0, full_recalc: 0},
```

Comment:
Shouldn't reason here for not displayed be "probability_selection" as we have positions: [{index: 0}, {index: 1}] available for both spocs?

Reply:
After walking through code, help clarify out_of_position as reason for not displaying spocs. In selectLayoutRender there can be only one unchosenspoc  with reason as probability_selection as it will use all the available positions. This is hard to understand by looking at code, may be a line or two explaining what's happening [here](https://github.com/mozilla/activity-stream/pull/4928/files#diff-91e34b3436d635509e2017c193c45ec1R114) and adding tests with multiple out_of_positions inside  spocsFill as reason will help.  
Here's gist with few tests that I used  to debug flow https://gist.github.com/punamdahiya/8daf45627380412ea48d9e174d2409c1

Reply:
The way `maybeInjectSpocs` works is that it iterates over all the SPOCS positions, and for each position, it checks the probability with a generated random number so that a SPOC will be injected if it meets the selection criterion; Otherwise, it will skip the SPOC injection for this position and move forward to the next position with the same SPOC.

In this case, for both two positions, it didn't meet the probability selection criterion for the first SPOC with the reason `probability_selection`. And the second SPOC didn't even get a chance to be selected due to the lack of positions. Therefore, its reason was set as `out_of_position`.

The same applies to the following cases as well.

NB: the current probability configuration would allow every check to pass, i.e. no `probability_selection` should arise. But it's subject to future changes that could be made on the server side.

Reply:
Good point, will add some comments to clarify it.